BaseWeapon/src/main/java/com/github/sef24sp4/baseweapon/BaseWeapon.java

```diff
@@ -0,0 +1,61 @@
+package com.github.sef24sp4.baseweapon;
+
+import com.github.sef24sp4.common.entities.ICollidableEntity;
+import com.github.sef24sp4.common.entities.IEntity;
+import com.github.sef24sp4.common.interfaces.IEntityManager;
+import com.github.sef24sp4.common.weapon.WeaponSPI;
+
+public class BaseWeapon implements WeaponSPI {
+	private int ammoCount;
+	private final int maxAmmoCount;
+	private long remainingCoolDownTicks = 0;
+	private long timeOfShoot = 0;
+	private final long presentTime = System.nanoTime();
+
+	public BaseWeapon() {
+		this.ammoCount = this.getAmmoCount();
+		this.maxAmmoCount = Integer.MAX_VALUE;
+		this.remainingCoolDownTicks = 0;
+		this.timeOfShoot = 0;
+	}
+
+	@Override
+	public boolean shoot(IEntityManager entityManager, IEntity shooter) {
+		if (this.getRemainingCoolDownTicks() < 0 && this.getAmmoCount() <= 0) {
+			return false;
+		} else if (this.getRemainingCoolDownTicks() > 0 && this.getAmmoCount() > 0) {
+			return true;
+		}
+
+		while (this.getRemainingCoolDownTicks() > 0 && this.getAmmoCount() > 0) {
+			this.ammoCount--;
+		}
+
+		this.remainingCoolDownTicks = this.getRemainingCoolDownTicks();
+
+		ProjectileControlSystem projectileControlSystem = new ProjectileControlSystem();
```

Comment:
Please store this as a field for the class. This avoids expensive call to constructor on every shoot.